### PR TITLE
Fix App does not automatically reconnect to HU after changing language

### LINKED ISSUE #1532: App does not automatically reconnect to HU after changing language.
### Bug Report
App does not automatically reconnect to HU after changing language.

### Preconditions
1. Connect the head unit to the phone via BT+USB
2. App supports multiple languages
3. SDLProxy version is 6.4.1

##### Reproduction Steps
1. Activate App
2. Change the language from English to Japanese on the HU

##### Expected Behavior
App automatically reconnects to HU after changing language.

##### Observed Behavior
The session closed right after changing language, but App does not automatically reconnect to HU.
※There is no problem with SDLProxy 6.3.1.

##### OS & Version Information
* iOS Version: 10.2
* SDL iOS Version: 6.4.1
* Testing Against: SDL Core 6.0.1 + HMI running on our development environment.

##### Occurrence
30%

##### Recovery Method
Restart the APP

##### Log
```
2020-01-07 11:46:20.593076+0800 VRAudioTestAppAdmin3[3085:1021763] accessoryd sessionResult = 0, useSocketInterace = 0, sessionID = 0x0	
2020-01-07 11:46:20.593317+0800 VRAudioTestAppAdmin3[3085:1021763] ERROR - /BuildRoot/Library/Caches/com.apple.xbs/Sources/ExternalAccessory/ExternalAccessory-400.40.1/EASession.m:-[EASession dealloc] - 212 unable to close session for _accessory=0x2822f84e0 and sessionID=65536	
2020-01-07 11:46:20.593538+0800 VRAudioTestAppAdmin3[3085:1021763] yangjun IAPSession createSession self.eaSession = (null)	
2020-01-07 11:46:20.596762+0800 VRAudioTestAppAdmin3[3085:1021763] accessoryd sessionResult = 0, useSocketInterace = 0, sessionID = 0x0	
2020-01-07 11:46:20.596954+0800 VRAudioTestAppAdmin3[3085:1021763] ERROR - /BuildRoot/Library/Caches/com.apple.xbs/Sources/ExternalAccessory/ExternalAccessory-400.40.1/EASession.m:-[EASession dealloc] - 212 unable to close session for _accessory=0x2822f84e0 and sessionID=65536	
2020-01-07 11:46:20.597061+0800 VRAudioTestAppAdmin3[3085:1021763] yangjun IAPSession createSession self.eaSession = (null)	
2020-01-07 11:46:20.600088+0800 VRAudioTestAppAdmin3[3085:1021763] accessoryd sessionResult = 0, useSocketInterace = 0, sessionID = 0x0	
2020-01-07 11:46:20.600281+0800 VRAudioTestAppAdmin3[3085:1021763] ERROR - /BuildRoot/Library/Caches/com.apple.xbs/Sources/ExternalAccessory/ExternalAccessory-400.40.1/EASession.m:-[EASession dealloc] - 212 unable to close session for _accessory=0x2822f84e0 and sessionID=65536	
2020-01-07 11:46:20.600474+0800 VRAudioTestAppAdmin3[3085:1021763] yangjun IAPSession createSession self.eaSession = (null)	
```
[logs.txt](https://github.com/smartdevicelink/sdl_ios/files/4126412/logs.txt)

### PR DESCRIPTION
**Pull request by zhouxin627**
Fixes #1532 

This PR is **ready** for review.

### Risk 
This PR makes **no** API changes.

### Testing Plan
Covered by Unit tests.

### Summary
**Reason**
1. There are two threads (`ioStreamThread` and `lifecycleThread`)  have Proxy object
2. After HU changing language, Proxy's `UnregisterApp` is invoked
3. Then `lifecycleThread`'s `stopManager` is invoked
4. set `Proxy = nil` and then Proxy's reference count will reduce to `0` 
5. when Proxy's reference count reaches `0`，Proxy's `dealloc` method will be invoked

Normal case：
5.1.1 Proxy's `dealloc` is invoked in the `lifecycleThread`
5.1.2 set `ioStreamThread`'s cancel flag true
5.1.3 `ioStreamThread` exit
5.1.4 close Session
5.1.5 `ioStreamThread = nil`
5.1.6 create new session successfully
image

Abnormal case：
5.2.1 Proxy's `dealloc` is invoked in the `ioStreamThread`
5.2.2 set `ioStreamThread`'s cancel falg true
5.2.3 `ioStreamThread = nil`
5.2.4 `ioStreamThread` exit faild
5.2.5 create new session failed
image

**Solution**
Avoid `ioStreamThread` do expensive operations, so move the methods in the `dealloc` to the new method of `disconnectSession`, before set `Proxy = nil`, execute `disconnectSession`.
	
1. There are two threads (`ioStreamThread` and `lifecycleThread`)  have Proxy object	
2. After HU changing language, Proxy's `UnregisterApp` is invoked	
3. Then `lifecycleThread`'s `stopManager` is invoked	
4. Invoke Proxy's `disconnectSession` method in the `lifecycleThread`	
5. Set `ioStreamThread`'s cancel flag true	
6. `ioStreamThread` exit	
7. Close Session
8. `ioStreamThread = nil`
9. Set `Proxy = nil` and then Proxy's reference count will reduce to `0` 	
10. When Proxy's reference count reaches `0`，Proxy's `dealloc` method will be invoked	
11. Create new session successful	
image

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by joeljfischer** [comment]
@zhouxin627 Thanks for the PR. Have you determined if this PR or something similar is needed on Java_Suite?

**Comment by zhouxin627** [comment]
Hi @joeljfischer, I think there is no need to modify Java_Suite because it can work well.

**Comment by joeljfischer** [comment]
A few questions:

> There are two threads (ioStreamThread and lifecycleThread) have Proxy object

This doesn't seem to be true, only the lifecycle manager has a proxy object.

> Proxy's dealloc is invoked in the ioStreamThread

Where does this happen. `proxy = nil` only occurs in `SDLLifecycleManager.sdl_stopManager:` that I can find (7 on your "abnormal case" chart).

**Comment by zhouxin627** [comment]
Hi @joeljfischer 

> This doesn't seem to be true, only the lifecycle manager has a proxy object.

When the head unit sends data to the phone, the session will transfer data to Proxy, and this
transferring is handled in ioStreamThread.

> Where does this happen. `proxy = nil` only occurs in `SDLLifecycleManager.sdl_stopManager:` that I can find (7 on your "abnormal case" chart).

The function(`proxy = nil`) is indeed handled in `sdl_stopManager`,  but the iOS project manages resources with ARC. When ARC works, the time of `dealloc` and execution thread is based on the release function called by the system,  which is used in the case of the count is zero. 

The proxy object is owed by `ioStreamThread` and `LlifecycleThread`, so the last time releasing will probably happen in the `ioStreamThread`,  and this will cause `dealloc` to execute in `ioStreamThread`.

File: SmartDeviceLink/SDLMutableDataQueue.m [hunk 1]
```diff
@@ -10,6 +10,8 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+NSInteger const MAXDATACOUNT = 200;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMutableDataQueue.m]: The changes in this class have nothing to do with this issue. Please revert.

File: SmartDeviceLink/SDLProxy.h [hunk 2]
```diff
@@ -133,6 +133,11 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)putFileStream:(NSInputStream *)inputStream withRequest:(SDLPutFile *)putFileRPCRequest;
 
+/**
+ * disconnect session
+ */
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.h]:
```suggestion
/// Disconnects the current app session , including the security manager and primary transport.
```

File: SmartDeviceLink/SDLProxy.m [hunk 3]
```diff
@@ -174,7 +174,8 @@ + (SDLProxy *)tcpProxyWithListener:(id<SDLProxyListener>)delegate tcpIPAddress:(
     return ret;
 }
 
-- (void)dealloc {
+- (void)disconnectSession {
+    SDLLogD(@"Proxy disconnectSession current thread = %@", NSThread.currentThread);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.m]:
```suggestion
    SDLLogD(@"Disconnecting the proxy; stopping security manager and primary transport.);
```

**Comment by joeljfischer** [comment]
Hello @zhouxin627, are you considering this bug no longer as needing a fix? Are you planning on re-opening this PR?

**Comment by zhouxin627** [comment]
Hi @joeljfischer, yes, I want to re-opening this PR. It was closed automatically when I revert `NSInteger const MAXDATACOUNT = 200;`. Now, It is showing No commits yet and `Pushing new commits will allow the pull request to be re-opened.`. I'm investigating how to deal with it.

File: SmartDeviceLink/SDLMutableDataQueue.m (hunk 1 above)
Comment (zhouxin627) [reply SmartDeviceLink/SDLMutableDataQueue.m]: @joeljfischer done in 8737234

File: SmartDeviceLink/SDLProxy.h (hunk 2 above)
Comment (zhouxin627) [reply SmartDeviceLink/SDLProxy.h]: @joeljfischer done in 8737234

File: SmartDeviceLink/SDLProxy.m (hunk 3 above)
Comment (zhouxin627) [reply SmartDeviceLink/SDLProxy.m]: @joeljfischer done in 8737234

**Comment by theresalech** [comment]
@zhouxin627 can you please confirm if this PR is now ready for Livio's review?

**Comment by zhouxin627** [comment]
@theresalech Hi, this PR is now ready for review.

File: SmartDeviceLink/SDLProxy.h [hunk 4]
```diff
@@ -133,6 +133,9 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)putFileStream:(NSInputStream *)inputStream withRequest:(SDLPutFile *)putFileRPCRequest;
 
+/// Disconnects the current app session , including the security manager and primary transport.
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply SmartDeviceLink/SDLProxy.h:136]: @joeljfischer thanks, done.

**Pull request merged**